// syndicate-interface/lib/things/thing.js

var assert = require('assert')

var mentions = require('../mentions/extract')
var validate = require('../validate')

module.exports = Thing

function Thing() {
  // deleted stuff is hidden
  this.deleted = false
  // closed stuff has features disabled (like comments)
  this.closed = false
  // store all the dates in one place
  this.date = {
    created: new Date()
  }
}

/**
 * Extend another collection from this collection.
 */

Thing.extend = function (con) {
  con.set = Object.create(this.set)
}

/**
 * Set various properties on almost anything.
 */

Thing.set = {}

/**
 * Titles are not unique!
 */

Thing.set.title = function (thing, str) {
  thing.title = validate.title(str)
}

/**
 * We don't parse the markdown server side!
 * Make sure it's not too long!
 */

Thing.set.markdown = function* (thing, str) {
  thing.markdown = validate.markdown(str)
  thing.mentions = yield mentions
}

/**
 * Set a relation on the current object.
 */

Thing.set.relation = function (thing, relation, child) {
  assert(child)
  assert(child._id)
  assert(child.collection)

  thing[relation] = {
    _id: child._id,
    collection: child.collection
  }
}
