// syndicate-interface/lib/validate/index.js

var ObjectID = require('mongodb').ObjectID
var validator = require('validator')
var assert = require('http-assert')

var re = {
  oid: /^[0-9a-f]{24}$/i,
  slug: /^\w{3,30}$/i,
  twitter: /^@\w{1,15}$/i,
  name: /^[\w \-.'"()]{3,100}$/i,
  password: /^.{8,}$/,
}

exports.oid = function (str) {
  assert(str, 422, 'Invalid ID.')
  var c = str.constructor
  if (c && c.name === 'ObjectID') return str
  assert(re.oid.test(str), 422, 'Invalid ID.')
  return new ObjectID(str)
}

exports.slug = function (str) {
  assert(str && !re.oid.test(str) && re.slug.test(str), 422, 'Invalid slug.')
  return str.toLowerCase()
}

exports.name = function (str) {
  assert(str && !re.oid.test(str) && re.name.test(str), 422, 'Invalid name.')
  return str.replace(/\s+/g, ' ').trim()
}

exports.password = function (str) {
  assert(re.password.test(str), 422, 'Invalid password. Passwords must be at least 8 characters long.')
  return str
}

exports.email = function (str) {
  str = (str || '').toLowerCase()
  assert(str && validator.isEmail(str), 422, 'Invalid email address.')
  return str.trim().toLowerCase()
}

// TODO: make sure it's not too long
exports.markdown = function (str) {
  str = str || ''
  return str.trim()
}

// TODO: make sure it's not too long
exports.title = function (str) {
  str = str || ''
  return str.replace(/\s+/g, ' ').trim()
}
